test/raft/distributed_test.cc

```diff
@@ -804,5 +805,233 @@ future<> stop_server(raft_server<M>& s) {
     }
 }
 
-SEASTAR_THREAD_TEST_CASE(dummy_test) {
+
+struct RWRegister {
+    struct write { int32_t x; };
+    struct read {};
+
+    struct ack {};
+    struct ret { int32_t x; };
+
+    using state_t = int32_t;
+    using input_t = std::variant<write, read>;
+    using output_t = std::variant<ack, ret>;
+
+    static std::pair<state_t, output_t> delta(state_t current, input_t input) {
+        using res_t = std::pair<state_t, output_t>;
+
+        return std::visit(make_visitor(
+        [] (const write& w) -> res_t {
+            return {w.x, ack{}};
+        },
+        [&current] (const read&) -> res_t {
+            return {std::move(current), ret{current}};
+        }
+        ), input);
+    }
+
+    static const state_t init = 0;
+};
+
+namespace ser {
+    template <>
+    struct serializer<RWRegister::read> {
+        template <typename Output>
+        static void write(Output& buf, const RWRegister::read&) {};
+
+        template <typename Input>
+        static RWRegister::read read(Input& buf) { return RWRegister::read{}; }
+
+        template <typename Input>
+        static void skip(Input& buf) {}
+    };
+
+    template <>
+    struct serializer<RWRegister::write> {
+        template <typename Output>
+        static void write(Output& buf, const RWRegister::write& w) { serializer<int32_t>::write(buf, w.x); };
+
+        template <typename Input>
+        static RWRegister::write read(Input& buf) { return { serializer<int32_t>::read(buf) }; }
+
+        template <typename Input>
+        static void skip(Input& buf) { serializer<int32_t>::skip(buf); }
+    };
+}
+
+// Raft uses UUID 0 as special case.
+// Convert local 0-based integer id to raft +1 UUID
+static utils::UUID to_raft_uuid(size_t local_id) {
+    return utils::UUID{0, local_id + 1};
+}
+
+static raft::server_id to_raft_id(size_t local_id) {
+    return raft::server_id{to_raft_uuid(local_id)};
+}
+
+future<> basic_test() {
+    using state_t = typename RWRegister::state_t;
+    using output_t = typename RWRegister::output_t;
+
+    struct delivery_target {
+        delivery_queue<state_t>& _queue;
+        failure_detector& _fd;
+    };
+
+    // Before we show a raft server to others
+    // we must add its delivery queue and fd to this map.
+    std::unordered_map<raft::server_id, delivery_target> queues;
+    network<message_t<state_t>> net([&queues]
+            (raft::server_id src, raft::server_id dst, const message_t<state_t>& m) -> bool {
+        auto it = queues.find(dst);
+        assert(it != queues.end());
+
+        auto& [q, fd] = it->second;
+        fd.receive_heartbeat(src);
```

Comment:
May be I am missing it but do you send heartbeats or only use regular messages as indication that a nodes is alive?

Reply:
In `failure_detector::advance` we broadcast heartbeats to all known servers. This is implemented by sending a message with `nullopt` payload through the network in `create_server`:
```
    auto fd = seastar::make_shared<failure_detector>(
            [id, &net] (raft::server_id dst) {
        net.send(id, dst, std::nullopt);
    });
```

When we receive a message from another server: if it's `nullopt`, it's just a heartbeat, but if it's an engaged optional, it's both a heartbeat and RPC message.